### PR TITLE
Newchecker

### PR DESCRIPTION
**Pull request by Slowblitz**
Taking over #61 :
So as we discuss earlier I'll continue this PR, there is still some stuff to be done as describe below :

- [x] Fix the TODO in the AnDOChecker.py
- [x] add new datasets (One dataset per "rules" for each level)
- [x] Add new unit test that check regexp and overall functioning  (this one should be done first if we want to do it in a "test driven development" way).
- [x] Add the generate regexp function 
- [ ] add a proper Error handling 

Feel free to add task that I have maybe forgot =)

### CODE REVIEW

File: ando/AnDOChecker.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels.tsv',
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels.json',
+        ]
+    """
+    list_of_rules = list()
+    if file_ext_regexp is None:
+        list_of_rules = file_name_regexp
+    else:
+        if len(file_name_regexp) == 2:
+            for filename in file_name_regexp[:-1]:
```
[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:151]: So the file_name_regexp is not just a list of filenames, but can also contain the potential extensions. Shouldn't then the extension be used to build the rule itself instead of the file_ext_regexp?

File: ando/AnDOChecker.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                        error_list.append("Mandatory file not found for this rule : {}".
+                                          format(current_mandatoryfolder_rule))
+
+    # if there are no errors, the data set is valid!
+    valid = len(error_list) == 0
+
+    return valid, error_list
+
+
+def get_list_of_rules_with(file_name_regexp, file_ext_regexp=None):
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:110]: I think the variable names used here could be more explicit. Since this function is rather generic, maybe also `build_rule_regexp` would be a better name? And `file_name_regexp` can actually contain more than just that, as it's a list of regexp_sets (is it actually restricted to filenames?).

File: ando/AnDOChecker.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        # 1.check whether the "mandatory Folders" are present at this level
+        #
+        ###
+        if len(currentdepth_rules['mandatory_folders']) > 0:
+            # loop over rules, each rule corresponding to one mandatory folder
+            for current_mandatoryfolder_rule in (currentdepth_rules['mandatory_folders']):
+                # create the list of regexp
+                list_of_mandatory_folders = (
+                    get_list_of_rules_with(current_mandatoryfolder_rule))
+                for each_file in list_of_mandatory_folders:
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:58]: is this now dealing with mandatory folders or mandatory files? the variable names here are mixed...

File: ando/AnDOChecker.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                    dir_res = [re.compile(each_file).search(d) is None for d in dirs]
+                    if all(dir_res):
+                        error_list.append(
+                            "Mandatory folder not found for this rule : {}".format(current_mandatoryfolder_rule))
+        ###
+        # 2. check whether the rules are followed for the folder at this level ("authorized folders")
+        #
+        ###
+        folder = op.split(root)[1]
+        folder_res = [re.compile(x).search(folder) is None
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:68]: Based on the variable name `folder_res` it is not intuitive if True or False means that the validation was successful. Maybe there's a better name for this?

Comment (JuliaSprenger) [inline ando/AnDOChecker.py:68]: This kind of list comprehension is repeatedly used in the code. Should this be separated into a utility function, so it can also be documented better?

File: ando/AnDOChecker.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                      for x in get_list_of_rules_with(currentdepth_rules['authorized_folders'])]
+
+        # if none of the authorized rules is respected, raise an error
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:71]: This could be the case, e.g. for an empty repository. If we follow the BIDS validator in this case then also an empty folder should be valid, right?

File: ando/AnDOChecker.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            for filename in file_name_regexp:
+                for extension in file_ext_regexp:
+                    list_of_rules.append(filename + str(extension))
+
+    return list_of_rules
+
+
+def main():
+    """
+    Main file of the validator. uses other class methods for checking
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:164]:
```suggestion
    Main function of the validator. This coordinates the core AnDO functions for checking
```

File: ando/AnDOChecker.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    parser.add_argument('directory', help='Name of the directory to be checked')
+    args = parser.parse_args()
+
+    try:
+        directory = args.directory
+    except IndexError:
+        directory = '.'
+
+    if not directory:
+        print('Directory does not exist:', directory)
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:191]: Is this still `print` and `exit` based to be compatible with the online AnDO checker? Otherwise it might be better to use exceptions here...

File: ando/AnDOChecker.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        exit(1)
+    dataset_validity, error_list = is_valid(directory)
+    if dataset_validity:
+        print("Congratulations!\n" +
+              directory +
+              " respects the ephys-BIDS specifications")
```
Comment (JuliaSprenger) [inline ando/AnDOChecker.py:197]:
```suggestion
        print("Congratulations!\n"
            f"{directory} respects the ephys-BIDS specifications")
```

Comment (JuliaSprenger) [reply ando/AnDOChecker.py:197]: since this is a new package we can use the new Python string formatting features ;) (and also in the other prints below...)

File: ando/rulesStructured.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+currentdepth_rules['authorized_data_files'] = [['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys']]
+currentdepth_rules['authorized_metadata_files'] = [
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys'],
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels'],
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_contacts'],
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_probes'],
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_runs']
+    ]
+currentdepth_rules['mandatory_files'] = [
+    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys', [f'({AUTHORIZED_DATA_EXTENSIONS[0]}|{AUTHORIZED_DATA_EXTENSIONS[1]})']],
```
Comment (JuliaSprenger) [inline ando/rulesStructured.py:49]: this should also work and is a bit shorter....
```suggestion
    ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys', [f'({"|".join(AUTHORIZED_DATA_EXTENSIONS})']],
```

File: ando/tests/test_AnDOChecker.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    ##
+    # THE VALID ONES
+    ##
+    def test_expValid(self):
+        path = Path(dir_path) / "dataset" / "exp-valid"
+        self.assertEqual(CHK.is_valid(path)[0], True)
+
+    def test_expValidMultipleSession(self):
+        path = Path(dir_path) / "dataset" / "exp-validMultipleSession"
```
Comment (JuliaSprenger) [inline ando/tests/test_AnDOChecker.py:19]: CamelCase is not for pep 8, luckily this is only folder names and not code ;)

File: pytest.ini [hunk 11]
```diff
@@ -0,0 +1,5 @@
+# pytest.ini
+[pytest]
+minversion = 5.0
+addopts = -ra -q
+testpaths =
```
Comment (JuliaSprenger) [inline pytest.ini:5]: shouldn't there be some testpath defined here or is this line intentionally incomplete?

File: ando/AnDOChecker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py:151]: Yes you're right I should use the last part of the  list with the list of extension to build the rules.But it means that we are not using the 2nd parameter given in argument any more

File: ando/AnDOChecker.py (hunk 2 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py:110]: ```file_name_regexp``` Can be both filenames or foldernames actually so yes it's a good idea to rename it

File: ando/AnDOChecker.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        # print(depth,root,dirs,files)
+        ###
+        # extract rules for this level!
+        ###
+        currentdepth_rules = Rs.rules_set[depth]
+        ###
+        # 1.check whether the "mandatory Folders" are present at this level
+        #
+        ###
+        if len(currentdepth_rules['mandatory_folders']) > 0:
```
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py:52]: Except if we don't have any mandatory folder at some point

File: ando/AnDOChecker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:151]: I guess then we can actually remove it...

[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:151]: Makes also the docstring shorter ;)

File: ando/AnDOChecker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:52]: Then the list is empty and the loop will be skipped naturally anyway, no?

File: ando/AnDOChecker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py:151]: I can't remove it because we are using it when there is no extension in the first list

[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:151]: Ah, because that would require adding the authorized metadata extensions to all the metadata files in the rules... 
Which might be the conceptually cleaner solution as then the complete definition of all regexes would be in the rules file and no additional assumptions (like metadata file should have metadata information) would be coded in the validator. But also it would make the rule definition more complex again. What do you think?

File: ando/AnDOChecker.py (hunk 7 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py:191]: I don't know what Sylvain want for this one =)

File: ando/AnDOChecker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py:151]: Yes so , I've removed the 2nd argument of the function now all filenames and extension are in the same argument as : 
```[['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys',AUTHORIZED_DATA_EXTENSIONS]]```
with ```AUTHORIZED_METADATA_EXTENSIONS = [f'({"|".join([".tsv",".json"])})']```

File: ando/AnDOChecker.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    for ind, (root, dirs, files) in enumerate(os.walk(input_directory)):
+        # compute depth of the directory given as input
+        if ind == 0:
+            # remove ending / or \ if the input directory was given with it at the end
+            if root.endswith(os.path.sep):
+                root = root[:-1]
+            # count the number of / or \ in the directory name to estimate its "depth"
+            initial_depth = root.count(os.path.sep)
+        # estimate at which level we are
+        depth = root.count(os.path.sep) - initial_depth
```
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py]: Done

File: ando/AnDOChecker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Slowblitz) [reply ando/AnDOChecker.py]: done

File: ando/AnDOChecker.py (hunk 3 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py]: updated

File: ando/AnDOChecker.py (hunk 4 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py]: updated , named ```folders_err``` since it is true when the folder does not respect the rules

Comment (Slowblitz) [reply ando/AnDOChecker.py]: Done

File: ando/rulesStructured.py (hunk 9 above)
Comment (Slowblitz) [reply ando/rulesStructured.py]: Done

File: ando/AnDOChecker.py (hunk 8 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py]: Done

File: ando/AnDOChecker.py (hunk 2 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py]: Done

File: ando/AnDOChecker.py (hunk 4 above)
Comment (Slowblitz) [reply ando/AnDOChecker.py]: done

**Comment by JuliaSprenger** [comment]
Hi @Slowblitz I pushed some updates directly to your branch, don't be surprised when pushing the next time ;)

**Comment by Slowblitz** [comment]
Yes test are still failing =/ but thx for the update, so I have to pull now, so we don't get any conflict

**Comment by JuliaSprenger** [comment]
I also adjusted the datasets.
The failing tests are related to the old version of the generator in this branch and with the new version the tests are actually succeeding, see https://github.com/INT-NIT/AnDO/pull/62

**Comment by Slowblitz** [comment]
Ok so maybe it's time to delete the old one no ?

**Comment by JuliaSprenger** [comment]
Yes, I we can do that. I hope this won't mess up me rebasing the new generator on this branch :crossed_fingers:

**Comment by JuliaSprenger** [comment]
@Slowblitz so should I or do you want to do this?

**Comment by Slowblitz** [comment]
You can do it =)

**Comment by JuliaSprenger** [comment]
Tadaaa

**Comment by Slowblitz** [comment]
\o/ yeaaaa good job =)

File: ando/AnDOChecker.py (hunk 5 above)
Comment (JuliaSprenger) [reply ando/AnDOChecker.py:79]: We should add a test dataset that is empty.

**Comment by JuliaSprenger** [comment]
Regarding the last open checkbox for this PR, the error handling, I suggest to leave it as it is right now (the `is_valid` function returning a bool indicator and a list of error messages as this output is well defined and can be used by different interfaces (python scripts, webchecker,...)

**Comment by SylvainTakerkart** [comment]
OK, the tests are failing coz there's a problem with a relative import... I need to fix this!

**Comment by SylvainTakerkart** [comment]
@JuliaSprenger OK, tests pass! I let you merge if you want to take a look (or merge directly)...

File: ando/AnDOChecker.py (hunk 7 above)
Comment (SylvainTakerkart) [reply ando/AnDOChecker.py]: I had no special intention here... This looks like a copy-paste from some old code... (maybe from @Slowblitz ... :) ); but we probably can ignore this for now and deal with it later with the overall error handling...
